### PR TITLE
Ensure that Picture::RasterizeToImage destroys Dart persistent values on the UI thread

### PR DESCRIPTION
**Pull request by jason-simmons**
The DartPersistentValue used to hold the image callback is tied to a
Dart isolate.  Destructing the DartPersistentValue requires entering
the isolate and must be done on the UI thread.

Fixes https://github.com/flutter/flutter/issues/29379

### CODE REVIEW

File: lib/ui/painting/picture.cc [hunk 1]
```diff
@@ -84,52 +84,43 @@ Dart_Handle Picture::RasterizeToImage(sk_sp<SkPicture> picture,
...
-                return;
-              }
-
-              auto dart_image = CanvasImage::Create();
-              dart_image->set_image(
-                  {std::move(raster_image), std::move(unref_queue)});
-              auto* raw_dart_image = tonic::ToDart(std::move(dart_image));
-
-              // All done!
-              tonic::DartInvoke(image_callback->Get(), {raw_dart_image});
```
Comment (liyuqian) [inline lib/ui/painting/picture.cc]: I'd expect that the `std::unique_ptr` to be destructed here after that `image_callback` is no longer used, which means destruction in the ui thread. Is it not the case? @chinmaygarde

Comment (jason-simmons) [reply lib/ui/painting/picture.cc]: In the old implementation, the `std::unique_ptr` was being captured into a ref counted lambda by `MakeCopyable`.

The `MakeCopyable` lambda was then being passed through the GPU thread to the UI thread.  Either the GPU thread or the UI thread could drop the last reference to the lambda.  If the GPU thread destroyed the lambda, then the captured image_callback would be destructed on the GPU thread, causing a failure.

**Comment by eggfly** [comment]
> The DartPersistentValue used to hold the image callback is tied to a
> Dart isolate. Destructing the DartPersistentValue requires entering
> the isolate and must be done on the UI thread.
> 
> Fixes [flutter/flutter#29379](https://github.com/flutter/flutter/issues/29379)

I've faced the same problem related to this vm shutdown:  

"Unable to Enter Isolate :  Multiple mutators entering an isolate / Dart VM is shutting down" in dart_api_impl.cc from Dart_EnterIsolate() function.


And I digged some code related to dart_persistent_value.cc, dart_isolate_scope.cc, dart_api_impl.cc, but still don't know why "Destructing the DartPersistentValue requires entering the isolate and must be done on the UI thread", May I ask you why UI Thread?

**Comment by eggfly** [comment]
If I created DartPersistentValue's unique_ptr in io task runner:
`std::make_unique<tonic::DartPersistentValue>(tonic::DartPersistentValue* image_callback =   dart_state, raw_image_callback) // in io runner`

Then should I manually make sure that DartPersistentValue's unique_ptr is destructed in io thread or also ui thread? And not using the ref count logic inside MakeCopyable lambda.

**Pull request merged**